generators/Reporting.fs

```diff
@@ -0,0 +1,92 @@
+    | Some Status.All,           _                        -> true
+    | _ -> false
+    
+type SummaryTypes =
+    | Custom of string
+    | Unimplemented of string
+    | MissingData of string
+    | Deprecated of string
+    | UpToDate of string
+    | Outdated of string * string * string
+    
+let nameFromSummaryType = function
+    | Custom n -> n
+    | Unimplemented  n -> n
+    | MissingData n -> n
+    | Deprecated n -> n
+    | UpToDate n -> n
+    | Outdated (n,_,_) -> n
+    
+let private summarizeExercise options (parseCanonicalData':string -> CanonicalData) (exercise:Exercise)  =
+
+    match exercise with
+    | Exercise.Custom custom ->                 SummaryTypes.Custom custom.Name
+    | Exercise.Unimplemented unimplemented ->   SummaryTypes.Unimplemented unimplemented.Name
+    | Exercise.MissingData missingData ->       SummaryTypes.MissingData missingData.Name
+    | Exercise.Deprecated deprecated ->         SummaryTypes.Deprecated deprecated.Name
+    | Exercise.Generator generator ->
+        let cData = parseCanonicalData' generator.Name
+        
+        match generator.ReadVersion (), cData.Version with
+        | a,b when a.Equals b ->  UpToDate generator.Name
+        | a,b -> Outdated (generator.Name,a,b)
+        
+let listExercises options =
+    Log.Information(sprintf "Listing exercises with status %s..." (options.Status.Value.ToString()))
+    
+    let parseCanonicalData' = CanonicalData.parseCanonicalData options
+    
+    let allExercises = createExercises options
+    
+    let longestNameLength = allExercises |> List.map exerciseName |> List.map (fun e -> e.Length) |> List.max
+    let indentAfter (name:string) = String.replicate (longestNameLength+2-name.Length) " "
+    
+    let exerciseGroups = allExercises
+                            |> List.filter (filterByStatus options parseCanonicalData')
+                            |> List.map (summarizeExercise options parseCanonicalData')
+                            |> List.groupBy (fun x -> x.GetType())
+    
+    exerciseGroups |> List.iter (fun (groupType,group) ->
+        let description = match group.Head with
+                          | SummaryTypes.Custom _ -> "have customized tests"
+                          | SummaryTypes.Unimplemented _ -> "are missing test generator"
+                          | SummaryTypes.MissingData _ -> "are missing canonical data"
+                          | SummaryTypes.Deprecated _ -> "are deprecated"
+                          | SummaryTypes.UpToDate _ -> "are up to date"
+                          | SummaryTypes.Outdated _ -> "are outdated"
+                          
+        Log.Information("{num} exercises {description}:", group.Length, description)
+        
+        group |> List.iter (fun exercise -> 
+            match exercise with
+            | SummaryTypes.Outdated (name,oldVer,newVer) -> Log.Information(" {name}{indent}({oldVer} -> {newVer})", name, indentAfter name, oldVer, newVer)
```

Comment:
Could you remove the space before the exercise name in the logging statement?

Reply:
If you prefer, but I meant it to be a small amount of indentation to visually distinguish the exercise list under the category summary. For now I added a couple more spaces. Does that look better?
```
[10:45:16 INF] 20 exercises are outdated:
[10:45:16 INF]    allergies                  (1.1.0 -> 1.2.0)
[10:45:16 INF]    alphametics                (1.2.0 -> 1.3.0)
[10:45:16 INF]    anagram                    (1.3.0 -> 1.4.0)
[10:45:16 INF]    atbash-cipher              (1.1.0 -> 1.2.0)
[10:45:16 INF]    book-store                 (1.3.0 -> 1.4.0)
[10:45:16 INF]    bracket-push               (1.3.0 -> 1.4.0)
(...)
```

Reply:
I just pushed up the single-space version, applied to both `-s` and `-e`. It'd be easy to change later if we want.

Reply:
For spacing things like this I don't have a strong preference one way over another, but I do have a strong preference for consistency :)